docs: adding dynamic channel page (#1959)
asyncapi/website: pages/docs/concepts/asyncapi-document/dynamic-channel-names.md

---

> title: Dynamic Channel Names

Reviewer: We talk about dynamic addresses. In previous versions, the address was also the name of the channel, but in v3, as I mentioned in https://github.com/asyncapi/website/pull/1959#discussion_r1265651267, it's moved into the `address` field.

This should be changed across the whole doc. 
```suggestion
title: Dynamic Channel Addresses
```

Or 

```suggestion
title: Dynamic Channel Address
```

Not sure what sounds better.

weight: 80
---

> Dynamic channel names, also known as channel address expressions, are a key feature of AsyncAPI. Dynamic channel names in AsyncAPI are pivotal for enhancing the flexibility, interoperability, reusability, and standardization of asynchronous APIs. They allow for variable usage in channel names that can be adapted to multiple APIs, used across different protocols, and reused across channels, thus preventing redundancy and maintaining consistency. This adaptability to different APIs aids in creating a unified view of the API, making it easier to comprehend and utilize. Dynamic channel names contribute to code and documentation generation and event management tooling, thereby improving APIs' overall efficiency and usability.

Reviewer: nitpick: I had to search for the term `pivotal` 😅 Not sure if you could replace by a more known one. Anyway, discard my suggestion if the word is well used out there and I'm the only one with the issue 😆

Reviewer 2: sure, we can use a synonym that seems less jargony such as `essential` or other ✌🏽

> Dynamic channel names, also known as channel address expressions, are a key feature of AsyncAPI. Dynamic channel names in AsyncAPI are pivotal for enhancing the flexibility, interoperability, reusability, and standardization of asynchronous APIs. They allow for variable usage in channel names that can be adapted to multiple APIs, used across different protocols, and reused across channels, thus preventing redundancy and maintaining consistency. This adaptability to different APIs aids in creating a unified view of the API, making it easier to comprehend and utilize. Dynamic channel names contribute to code and documentation generation and event management tooling, thereby improving APIs' overall efficiency and usability.

Reviewer: > Dynamic channel names, also known as channel address expressions, are a key feature of AsyncAPI. Dynamic channel names in AsyncAPI are pivotal for enhancing the flexibility, interoperability, reusability, and standardization of asynchronous APIs. They allow for variable usage in channel names that can be adapted to multiple APIs, used across different protocols, and reused across channels, thus preventing redundancy and maintaining consistency. This adaptability to different APIs aids in creating a unified view of the API, making it easier to comprehend and utilize. Dynamic channel names contribute to code and documentation generation and event management tooling, thereby improving APIs' overall efficiency and usability.

I like how this sentence sounds in overall, however I don't feel it provides much value to the user. Let me rephrase that. IMHO it adds too many good words that are not really important or related to the matter we talk about. I.e. there is only one (short) part of one sentence that explains the subject: ` They allow for variable usage in channel names`.
I would advocate for short, concise, and simpler documentation, without (much) ornaments. 

What do you think about focus on what it technically means (define dynamic values in the address AKA [Channel address expressions](https://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#channelAddressExpressions)) and tell by naming real use cases? I believe this will help users to understand what are talking about here.

```mermaid
graph TD
A[AsyncAPI]
B[Dynamic Channel Names]
D[Variable Usage in Channel Names]
E[Adapted to Multiple APIs and Protocols]
F[Reused Across Channels]

style B fill:#47BCEE,stroke:#47BCEE;

A --> B
B --> D
D --> E
D --> F

> ```

Reviewer: Will this graph let the user understand better what we are talking about in this doc? I feel this graph is not necessary.

## Parameter Context

> The Parameter Context clarifies the origin of parameters, ensuring consistent understanding across teams. It also enables efficient code generation, thereby accelerating development and enhancing the developer experience.

Reviewer: I would recommend to apply the same suggestion I wrote in https://github.com/asyncapi/website/pull/1959/files#r1281562860

> In AsyncAPI, the context of parameters in channel names often refers to where the parameter value comes from. For example, the parameter might be populated from the runtime event payload. This context can be noted in the AsyncAPI document, and code generators can use it to make the developer experience easier.

Reviewer: > to make the developer experience easier.

WDYT about adding the "how" to this statement? I mean explaining "How makes it easier?"?

```mermaid
flowchart TB
    subgraph "AsyncAPI Specification"
        A[Channel Names] --> B[Parameters]
        B --> C[Parameter Context]

        style C fill:#47BCEE,stroke:#47BCEE;

    end
    subgraph "Parameter Context"
        C --> D[Indicate the origin of parameter value]
        C --> E[Help code generators enhance developer experience]
    end
    subgraph "Parameter Sources"
        D --> F[Runtime Event Payload]
        D --> G[Other possible sources]

>     end

Reviewer: I'm not very sure this is providing a clear picture of what we talk about. In fact, I don't think we need a graph here. Same as https://github.com/asyncapi/website/pull/1959/files#r1281588954

```

Here is an example of parameter context:

```yml

> address: user/{userId}/signedup

Reviewer: Nitpick: for consistency with examples in our reference docs:

```suggestion
address: 'user/{userId}/signedup'
```

parameters:
  userId:
    description: Id of the user
```

## Reusing Parameters

> An important thing about parameters is that they can be reused. If there is another event, for example, a `UserUpdated` event, which also requires the userId, the parameter can be reused like the following example:

Reviewer: Applying our ubiquitous language, `message` is better than `event` here.

```suggestion
An important thing about parameters is that they can be reused. If there is another message, for example, a `UserUpdated` message, which also requires the userId, the parameter can be reused like the following example:
```

```yml
channels:
  userSignedUp:

>     address: user/{userId}/signedup

Reviewer: Nitpick: for consistency with examples in our reference docs:

```suggestion
    address: 'user/{userId}/signedup'
```

    parameters:
      userId:
        description: Id of the user.

>         location: $message.payload#/userid

Reviewer: Nitpick: for consistency with examples in our reference docs:
```suggestion
        location: '$message.payload#/userId'
```

Reviewer: I would also add the message object to the channel so you can add the `userid` property on it. I believe this will also help users to understand from where the value comes.

>     parameters:
>       userId:
>         description: Id of the user.
>         location: $message.payload#/UserUpdated      

Reviewer: This is duplicated. Did you validate this document?

```suggestion
```

Author: I have got the example from here.
![image](https://github.com/asyncapi/website/assets/14244685/56db3c7c-25c2-493e-b8f5-7c4cae15b92b)

However, now I have updated the example as per your other review.

operations: 
  userSignedUp:
    action: receive
    channel: 
      $ref: '#/channels/userSignedUp'  
```

> In this code, the previously defined userId parameter is reused to form the channel name for `UserUpdated` event.

Reviewer: Nitpick: for keeping ubiquitous language in place

```suggestion
In this AsyncAPI document, the previously defined userId parameter is reused to form the channel name for `UserUpdated` event.
```

Feel free to simplify to `document` if you prefer.
